File: examples/java/testing/examples/partial-failures.java
import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;
import software.amazon.awssdk.services.lambda.model.OperationStatus;
import software.amazon.lambda.durable.DurableContext;
import software.amazon.lambda.durable.model.ExecutionStatus;
import software.amazon.lambda.durable.testing.LocalDurableTestRunner;

class PartialFailuresTest {

    @Test
    void recordsWhichStepsSucceededBeforeFailure() {
        var runner = LocalDurableTestRunner.create(
            Void.class,
            (input, context) -> {
                context.step("step-1", String.class, ctx -> "ok");
                context.step("step-2", String.class, ctx -> "ok");
                context.step("step-3", String.class, ctx -> {
                    throw new RuntimeException("step-3 failed");
                });
                return null;
            }
        );

        var result = runner.runUntilComplete(null);

        assertEquals(ExecutionStatus.FAILED, result.getStatus());

        assertEquals(OperationStatus.SUCCEEDED, result.getOperation("step-1").getStatus());
        assertEquals(OperationStatus.SUCCEEDED, result.getOperation("step-2").getStatus());
        assertTrue(result.getError().isPresent());
    }
}

File: examples/csharp/testing/examples/partial-failures.cs
using System.Linq;
using Amazon.Lambda.DurableExecution;
using Amazon.Lambda.DurableExecution.Testing;
using Xunit;

public class PartialFailuresTest
{
    private static async Task<string> Workflow(object input, IDurableContext ctx)
    {
        await ctx.StepAsync(async (_, _) => "ok", name: "step-1");
        await ctx.StepAsync(async (_, _) => "ok", name: "step-2");
        return await ctx.StepAsync<string>(
            async (_, _) => throw new InvalidOperationException("step-3 failed"),
            name: "step-3");
    }

    [Fact]
    public async Task RecordsWhichStepsSucceededBeforeFailure()
    {
        await using var runner = new DurableTestRunner<object, string>(Workflow);

        TestResult<string> result = await runner.RunAsync(new object());

        Assert.True(result.IsFailed);

        Assert.Equal(OperationStatus.Succeeded, result.GetStep("step-1").Status);
        Assert.Equal(OperationStatus.Succeeded, result.GetStep("step-2").Status);
        Assert.NotNull(result.Error);
    }
}
